File: dotnet/Allmark/Block/ContentRule.cs
namespace Allmark.Block;

using Allmark.Types;

public static class ContentRule
{
    public static BlockRule Create()
    {
        return new BlockRule
        {
            Name = "content",
            TestStart = TestStart,
            TestContinue = (_, _) => false,
        };
    }

    private static bool TestStart(BlockParserState state, MarkdownNode parent, int endOfLine)
    {
        var content = state.Src.Substring(state.I, endOfLine - state.I) + Utils.GetLineEnding(state, endOfLine);
        if (parent.AcceptsContent)
        {
            if (state.HasBlankLine)
            {
                state.HasBlankLine = false;
            }
            else
            {
                parent.Content += new string(' ', state.Indent);
            }
        }
        else
        {
            parent.Content += state.Spaces;
            state.Spaces = "";
        }
        parent.Content += content;
        return true;
    }
}

File: web/src/block/contentRule.ts
import type BlockParserState from "../types/BlockParserState";
import type BlockRule from "../types/BlockRule";
import type MarkdownNode from "../types/MarkdownNode";
import getLineEnding from "../utils/getLineEnding";

const rule: BlockRule = {
	name: "content",
	testStart,
	testContinue: () => false,
};
export default rule;

function testStart(state: BlockParserState, parent: MarkdownNode, endOfLine: number) {
	let content = state.src.substring(state.i, endOfLine) + getLineEnding(state, endOfLine);
	if (parent.acceptsContent) {
		if (state.hasBlankLine) {
			state.hasBlankLine = false;
		} else {
			parent.content += " ".repeat(state.indent);
		}
	} else {
		parent.content += state.spaces;
		state.spaces = "";
	}
	parent.content += content;
	return true;
}
